File: src/scripts/backfill-embeddings.ts
import { MongoClient, ObjectId } from "mongodb";
import dotenv from "dotenv";
import { config } from "../config.js";
import { webCollections } from "../web/collections.js";
import { generateEmbeddingBatch } from "../web/llm/embeddings.js";

dotenv.config();

const BATCH_SIZE = 50;

const BEHAVIOR_TAGS = ["Aggressive", "Conservative", "LateNight", "CardCounterWatch", "PromoSeeker"] as const;
const GAME_TYPES = ["Baccarat", "Blackjack", "Roulette", "SicBo", "DragonTiger", "PokerRoom"] as const;

/** Pick `count` random items from an array without repeating. */
function randomSample<T>(arr: readonly T[], min: number, max: number): T[] {
  const count = min + Math.floor(Math.random() * (max - min + 1));
  const shuffled = [...arr].sort(() => Math.random() - 0.5);
  return shuffled.slice(0, Math.min(count, arr.length));
}

/**
 * Batch embed via LiteLLM gateway → local TEI (voyage-4-nano, 1024 dim).
 * Fails hard if the gateway returns an all-zero result (interpreted as
 * degraded state from the shared client) — this keeps backfill data quality
 * from silently corrupting.
 */
async function generateEmbeddingsBatch(texts: string[]): Promise<number[][]> {
  if (!config.llm.apiKey) {
    throw new Error("LITELLM_API_KEY is missing — please configure LLM gateway credentials.");
  }
  const vectors = await generateEmbeddingBatch(texts, "document");
  const anyZero = vectors.some((v) => v.every((x) => x === 0));
  if (anyZero) {
    throw new Error("LLM gateway returned zero-vector(s) — check LiteLLM/TEI health.");
  }
  return vectors;
}

async function backfillOffers(db: any) {
  console.log("Starting backfill for offer_catalog...");
  const collection = db.collection(webCollections.offers);
  const offers = await collection.find({}).toArray();
  console.log(`Found ${offers.length} offers to process.`);

  for (let i = 0; i < offers.length; i += BATCH_SIZE) {
    // Local TEI has no RPM limit — batch pacing is unnecessary.
    const batch = offers.slice(i, i + BATCH_SIZE);
    const offerTypeZh: Record<string, string> = {
      HotelRoom: "酒店禮遇",
      MusicShowTicket: "娛樂票券",
      PointsLimitedTime: "限時積分兌換",
      FNBVoucher: "餐飲禮券",
      CashRebate: "現金回扣",
      TransportVoucher: "專車接送禮券",
    };
    const texts = batch.map((o: any) => {
      const games = (o.targetGameTypes as string[] ?? []).join("、") || "所有遊戲";
      const rules = (o.eligibilityRules as string[] ?? []).join("；") || "無特定條件";
      const typeZh = offerTypeZh[o.offerType as string] ?? o.offerType;
      return (
        `${o.title}。` +
        `${o.description} ` +
        `優惠類型：${typeZh}。` +
        `目標遊戲：${games}。` +
        `適用條件：${rules}。`
      );
    });
    
    console.log(`Processing offers batch ${i / BATCH_SIZE + 1}...`);
    const embeddings = await generateEmbeddingsBatch(texts);

    for (let j = 0; j < batch.length; j++) {
      await collection.updateOne(
        { _id: batch[j]._id },
        { $set: { offerEmbedding: embeddings[j], updatedAt: new Date() } }
      );
    }
  }
  console.log("Offer catalog backfill complete.");
}

/**
 * Ensure every patron has at least one active session in patron_table_sessions.
 * For patrons without any active session, a synthetic session is inserted so
 * that behaviorTags are always available for embedding generation.
 */
async function ensurePatronSessions(db: any, patrons: any[]): Promise<Map<string, string[]>> {
  console.log("Ensuring all patrons have an active session...");
  const sessionsCollection = db.collection(webCollections.sessions);
  const tablesCollection = db.collection(webCollections.tables);

  // Load all active sessions and index by patronId
  const activeSessions = await sessionsCollection
    .find({ isActive: true }, { projection: { patronId: 1, behaviorTags: 1 } })
    .toArray();

  const sessionMap = new Map<string, string[]>();
  for (const s of activeSessions) {
    sessionMap.set(s.patronId, s.behaviorTags ?? []);
  }

  // Find patrons that have no active session
  const missingPatronIds = patrons
    .map((p: any) => p.patronId)
    .filter((id: string) => !sessionMap.has(id));

  if (missingPatronIds.length === 0) {
    console.log("All patrons already have an active session.");
    return sessionMap;
  }

  console.log(`Generating synthetic sessions for ${missingPatronIds.length} patrons...`);

  // Fetch a table to assign — fall back to a placeholder tableId if none exist
  const anyTable = await tablesCollection.findOne({}, { projection: { tableId: 1 } });
  const fallbackTableId = anyTable?.tableId ?? "T-0001";

  const now = new Date();
  const newSessions = missingPatronIds.map((patronId: string) => {
    const tags = randomSample(BEHAVIOR_TAGS, 1, 2);
    const seatedAt = new Date(now.getTime() - Math.random() * 3 * 60 * 60 * 1000); // up to 3h ago
    sessionMap.set(patronId, tags);
    return {
      _id: new ObjectId(),
      patronId,
      tableId: fallbackTableId,
      seatedAt,
      lastActionAt: new Date(seatedAt.getTime() + Math.random() * (now.getTime() - seatedAt.getTime())),
      sessionBetAmount: Math.floor(200 + Math.random() * 9800),
      currentStackEstimate: Math.floor(500 + Math.random() * 49500),
      behaviorTags: tags,
      isActive: true,
    };
  });

  await sessionsCollection.insertMany(newSessions);
  console.log(`Inserted ${newSessions.length} synthetic sessions.`);

  return sessionMap;
}

async function backfillPatrons(db: any) {
  console.log("Starting backfill for patron_profiles...");
  const collection = db.collection(webCollections.patrons);
  const patrons = await collection.find({}).toArray();
  console.log(`Found ${patrons.length} patrons to process.`);

  // Build a patronId -> behaviorTags map, inserting synthetic sessions where needed
  const sessionMap = await ensurePatronSessions(db, patrons);

  for (let i = 0; i < patrons.length; i += BATCH_SIZE) {
    // Local TEI has no RPM limit — batch pacing is unnecessary.
    const batch = patrons.slice(i, i + BATCH_SIZE);

    const regionZh: Record<string, string> = {
      Macau: "澳門本地",
      HongKong: "香港",
      Guangdong: "廣東省",
      OtherGBA: "大灣區",
      Taiwan: "台灣",
      International: "海外國際",
    };
    // Build a rich preference text that covers tier, games, ADT, points, risk flags, region and behavior
    const texts = batch.map((p: any) => {
      const games = (p.preferredGames as string[] ?? []).join(", ") || "unknown";
      const riskFlags = (p.riskFlags as string[] ?? []).filter((f: string) => f !== "None").join(", ") || "none";
      const behaviorTags = (sessionMap.get(p.patronId) ?? []).join(", ") || "unknown";
      const region = regionZh[p.region as string] ?? (p.region as string) ?? "不詳";
      return (
        `Patron tier ${p.tier}, ` +
        `ADT ${p.adt}, ` +
        `prefers games: ${games}, ` +
        `points balance ${p.pointsBalance ?? 0}, ` +
        `risk flags: ${riskFlags}, ` +
        `behavior: ${behaviorTags}, ` +
        `region: ${region}`
      );
    });

    console.log(`Processing patrons batch ${i / BATCH_SIZE + 1}...`);
    const embeddings = await generateEmbeddingsBatch(texts);

    for (let j = 0; j < batch.length; j++) {
      await collection.updateOne(
        { _id: batch[j]._id },
        { $set: { preferenceEmbedding: embeddings[j], updatedAt: new Date() } }
      );
    }
  }
  console.log("Patron profiles backfill complete.");
}

async function run() {
  if (!config.mongodbUri) {
    throw new Error("MONGODB_URI is missing");
  }

  const client = new MongoClient(config.mongodbUri);
  try {
    await client.connect();
    const db = client.db(config.databaseName);

    await backfillOffers(db);
    await backfillPatrons(db);

    console.log("ALL BACKFILLS COMPLETED SUCCESSFULLY.");
  } catch (err) {
    console.error("Backfill failed:", err);
    process.exit(1);
  } finally {
    await client.close();
  }
}

run();

File: server/app/scripts/backfill_embeddings.py
"""Batch-embed ``offer_catalog`` + ``patron_profiles`` via voyage-4-nano.

Port of ``src/scripts/backfill-embeddings.ts``. Uses ``BATCH_SIZE=50`` and
``generate_embedding_batch(texts, "document")``. Fails hard if any returned
vector is all-zero (means voyage local failed — silent zero fallback would
otherwise corrupt data).

Additional behaviour: patrons that lack an active session get a synthetic
``patron_table_sessions`` row inserted so ``behaviorTags`` are always
available when composing the embedding text.

Run with::

    uv run python -m app.scripts.backfill_embeddings
"""

from __future__ import annotations

import asyncio
import random
from datetime import datetime, timedelta, timezone
from typing import Any

from app.collections import web_collections as cols
from app.db import close_client, get_db
from app.llm.embeddings import generate_embedding_batch

BATCH_SIZE = 50

_BEHAVIOR_TAGS = ("Aggressive", "Conservative", "LateNight", "CardCounterWatch", "PromoSeeker")

_OFFER_TYPE_ZH: dict[str, str] = {
    "HotelRoom": "酒店禮遇",
    "MusicShowTicket": "娛樂票券",
    "PointsLimitedTime": "限時積分兌換",
    "FNBVoucher": "餐飲禮券",
    "CashRebate": "現金回扣",
    "TransportVoucher": "專車接送禮券",
}

_REGION_ZH: dict[str, str] = {
    "Macau": "澳門本地",
    "HongKong": "香港",
    "Guangdong": "廣東省",
    "OtherGBA": "大灣區",
    "Taiwan": "台灣",
    "International": "海外國際",
}


def _random_sample(seq, min_count: int, max_count: int) -> list[Any]:
    count = min_count + int(random.random() * (max_count - min_count + 1))
    count = min(count, len(seq))
    return random.sample(list(seq), count)


async def _generate_embeddings_batch(texts: list[str]) -> list[list[float]]:
    """Batch embed. Fails hard on zero-vectors (voyage local unavailable)."""
    vectors = await generate_embedding_batch(texts, "document")
    if any(all(x == 0 for x in v) for v in vectors):
        raise RuntimeError(
            "voyage embed returned zero-vector(s) — check voyage local runtime."
        )
    return vectors


async def _backfill_offers(db: Any) -> None:
    print("Starting backfill for offer_catalog...")
    coll = db[cols.offers]
    offers = await coll.find({}).to_list(length=None)
    print(f"Found {len(offers)} offers to process.")

    for i in range(0, len(offers), BATCH_SIZE):
        batch = offers[i : i + BATCH_SIZE]
        texts: list[str] = []
        for o in batch:
            games = "、".join(o.get("targetGameTypes") or []) or "所有遊戲"
            rules = "；".join(o.get("eligibilityRules") or []) or "無特定條件"
            type_zh = _OFFER_TYPE_ZH.get(str(o.get("offerType")), str(o.get("offerType")))
            texts.append(
                f"{o.get('title')}。"
                f"{o.get('description')} "
                f"優惠類型：{type_zh}。"
                f"目標遊戲：{games}。"
                f"適用條件：{rules}。"
            )

        print(f"Processing offers batch {i // BATCH_SIZE + 1}...")
        embeddings = await _generate_embeddings_batch(texts)

        now = datetime.now(timezone.utc)
        for offer, embedding in zip(batch, embeddings, strict=False):
            await coll.update_one(
                {"_id": offer["_id"]},
                {"$set": {"offerEmbedding": embedding, "updatedAt": now}},
            )
    print("Offer catalog backfill complete.")


async def _ensure_patron_sessions(
    db: Any, patrons: list[dict[str, Any]]
) -> dict[str, list[str]]:
    """Return a ``patronId -> behaviorTags`` map. Insert synthetic sessions
    for any patron that lacks an active one so the embedding text can always
    include behavior tags."""
    print("Ensuring all patrons have an active session...")
    sessions_coll = db[cols.sessions]
    tables_coll = db[cols.tables]

    active_sessions = await (
        sessions_coll.find(
            {"isActive": True},
            projection={"patronId": 1, "behaviorTags": 1},
        ).to_list(length=None)
    )

    session_map: dict[str, list[str]] = {}
    for s in active_sessions:
        session_map[s["patronId"]] = s.get("behaviorTags") or []

    missing_patron_ids = [
        p["patronId"] for p in patrons if p["patronId"] not in session_map
    ]

    if not missing_patron_ids:
        print("All patrons already have an active session.")
        return session_map

    print(f"Generating synthetic sessions for {len(missing_patron_ids)} patrons...")

    any_table = await tables_coll.find_one({}, projection={"tableId": 1})
    fallback_table_id = (any_table or {}).get("tableId") or "T-0001"

    now = datetime.now(timezone.utc)
    new_sessions: list[dict[str, Any]] = []
    for patron_id in missing_patron_ids:
        tags = _random_sample(_BEHAVIOR_TAGS, 1, 2)
        seated_at = now - timedelta(seconds=random.random() * 3 * 60 * 60)
        last_action_at = seated_at + timedelta(
            seconds=random.random() * max((now - seated_at).total_seconds(), 0)
        )
        session_map[patron_id] = tags
        new_sessions.append(
            {
                "patronId": patron_id,
                "tableId": fallback_table_id,
                "seatedAt": seated_at,
                "lastActionAt": last_action_at,
                "sessionBetAmount": int(200 + random.random() * 9800),
                "currentStackEstimate": int(500 + random.random() * 49500),
                "behaviorTags": tags,
                "isActive": True,
            }
        )

    if new_sessions:
        await sessions_coll.insert_many(new_sessions)
    print(f"Inserted {len(new_sessions)} synthetic sessions.")

    return session_map


async def _backfill_patrons(db: Any) -> None:
    print("Starting backfill for patron_profiles...")
    coll = db[cols.patrons]
    patrons = await coll.find({}).to_list(length=None)
    print(f"Found {len(patrons)} patrons to process.")

    session_map = await _ensure_patron_sessions(db, patrons)

    for i in range(0, len(patrons), BATCH_SIZE):
        batch = patrons[i : i + BATCH_SIZE]
        texts: list[str] = []
        for p in batch:
            games = ", ".join(p.get("preferredGames") or []) or "unknown"
            risk_flags = (
                ", ".join(f for f in (p.get("riskFlags") or []) if f != "None")
                or "none"
            )
            behavior_tags = ", ".join(session_map.get(p["patronId"], [])) or "unknown"
            region = _REGION_ZH.get(str(p.get("region")), str(p.get("region") or "不詳"))
            texts.append(
                f"Patron tier {p.get('tier')}, "
                f"ADT {p.get('adt')}, "
                f"prefers games: {games}, "
                f"points balance {p.get('pointsBalance', 0)}, "
                f"risk flags: {risk_flags}, "
                f"behavior: {behavior_tags}, "
                f"region: {region}"
            )

        print(f"Processing patrons batch {i // BATCH_SIZE + 1}...")
        embeddings = await _generate_embeddings_batch(texts)

        now = datetime.now(timezone.utc)
        for patron, embedding in zip(batch, embeddings, strict=False):
            await coll.update_one(
                {"_id": patron["_id"]},
                {"$set": {"preferenceEmbedding": embedding, "updatedAt": now}},
            )
    print("Patron profiles backfill complete.")


async def main() -> None:
    await _backfill_offers(get_db())
    await _backfill_patrons(get_db())
    print("ALL BACKFILLS COMPLETED SUCCESSFULLY.")


async def _entrypoint() -> None:
    try:
        await main()
    finally:
        await close_client()


if __name__ == "__main__":
    asyncio.run(_entrypoint())
